### PR TITLE
Cascade deletes to datasets when deleting experiments

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/13741?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13741/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13741
```

</p>
</details>

### Related Issues/PRs

 #10699

### What changes are proposed in this pull request?

Following up on the original fix by @chilir in #11966

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Before:

Repro steps:
1. Run this script (or any script to log a dataset)

```python
import pandas as pd

import mlflow

mlflow.set_tracking_uri("http://localhost:5000")
mlflow.set_experiment("datasets")

dataset_source_url = "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/winequality-red.csv"
raw_data = pd.read_csv(dataset_source_url, delimiter=";")

# Create an instance of a PandasDataset
dataset = mlflow.data.from_pandas(
    raw_data, source=dataset_source_url, name="wine quality - red", targets="quality"
)

# Log the Dataset to an MLflow run by using the `log_input` API
with mlflow.start_run() as run:
    mlflow.log_input(dataset)
```

2. Delete the "datasets" experiment from UI

3. Run the following in command line:

```
$ MLFLOW_TRACKING_URI=http://localhost:5000 mlflow gc --backend-store-uri=postgresql://postgres:password@localhost:5432/mlflow

...

mlflow.exceptions.MlflowException: (psycopg2.errors.ForeignKeyViolation) update or delete on table "experiments" violates foreign key constraint "datasets_experiment_id_fkey" on table "datasets"
DETAIL:  Key (experiment_id)=(1) is still referenced from table "datasets".

[SQL: DELETE FROM experiments WHERE experiments.experiment_id = %(experiment_id)s]
[parameters: {'experiment_id': 1}]
```

This flow works after the migration.

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixed a bug where experiment deletion would not cascade to datasets, preventing experiments with datasets from being deleted. Please run `mlflow db upgrade <DB_URI>` to utilize the fix!

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/test_cli.py
Comment (TomeHirata) [inline tests/test_cli.py:606]: question: do we need to use the real dataset for this test? I wonder if this practice increases the text execution time.

Comment (daniellok-db) [reply tests/test_cli.py:606]: Oh true, we don't, let me update

File: tests/db/schemas/mssql.sql
Comment (daniellok-db) [inline tests/db/schemas/mssql.sql:182]: not sure why this changed, i think maybe we missed out on running `./tests/db/update_schemas.sh` in #13176?

**Comment by daniellok-db** [comment]
cc @harupy what is the `test_store_generated_schema_matches_base` test for? it seems to be failing because we're removing a foreign key, but we do add it back so i'm not sure what's going on there

File: tests/db/schemas/mssql.sql
Comment (harupy) [reply tests/db/schemas/mssql.sql:182]: I ran `update_schemas.sh`:

https://github.com/mlflow/mlflow/actions/runs/11790377345/job/32840716342?pr=13740

It looks like we only check `tests/resources/db/latest_schema.sql`.

**Comment by harupy** [comment]
I'm not sure what `test_store_generated_schema_matches_base` is for but it looks like it validate there is no diff between the base schema and migrated schema.

**Comment by daniellok-db** [comment]
Found the issue, needed to edit the ORM definitions in `mlflow/store/tracking/dbmodels/models.py`. Let me add a comment to the test.

File: tests/db/schemas/mssql.sql
Comment (harupy) [reply tests/db/schemas/mssql.sql:182]: Filed https://github.com/mlflow/mlflow/pull/13761

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:621]:
```suggestion
    experiments = store.search_experiments(view_type=ViewType.ALL)
```

can we remove `filter_string`?

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:625]:
```suggestion
    experiments = store.search_experiments(view_type=ViewType.ALL)
```

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:624]:
```suggestion
    subprocess.check_output([sys.executable, "-m", "mlflow", "gc", "--backend-store-uri", sqlite_store[1]])
```

File: mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py
Comment (harupy) [inline mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:1]: `0584bdc529eb_add_cascading_deletion_to_datasets_from_.py` -> mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_experiments.py

File: mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py
Comment (harupy) [inline mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: Is it possible to name this constraint?

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:624]:
```suggestion
    subprocess.check_call(["mlflow", "gc", "--backend-store-uri", sqlite_store[1]])
```

because we don't use output.

File: mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py
Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: i think it's possible but i'm not sure of the implications. maybe we can just standardize on `fk_datasets_experiment_id_experiments"

Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: to be safe i feel like it's probably best to do less changes, but i don't have a lot of knowledge about DB stuff. if it's not a problem for the names to change then i'll update the PR!

Comment (harupy) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: Makes sense!

File: tests/store/tracking/test_sqlalchemy_store_schema.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store_schema.py:113]: Can we include this in the assert error message? That's more noticeable.

File: mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py
Comment (harupy) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: This reminds me of https://github.com/mlflow/mlflow/pull/11852#discussion_r1583428781:

> Can we name the foreign key please? Otherwise MSSQL will generate a generated unique uuid based name that forces additional processing steps for fkey migrations or modifications

Comment (harupy) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py:56]: can we try naming the constraint? `tests/db/check_migration.py` should pass if renaming doesn't work or is not allowed.

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/13741#discussion_r1837525953 is addressed!

File: mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py
Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py]: it looks like it works okay, how do i run that script in a docker container?

Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py]: or can we just wait for CI

Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py]: hmm it's quite weird, running `./tests/db/compose.sh run migration-postgresql` i get the following error during the pre-migration check:

```
mlflow.exceptions.MlflowException: Detected out-of-date database schema (found version 0584bdc529eb, but expected f5a4f2784254).
```

it seems the database is already migrated (`0584bdc529eb` is the revision in this PR), maybe some caching thing on my end? i'm not sure if that's the right command to run

Comment (harupy) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py]: Interesting. Let me try.

Comment (daniellok-db) [reply mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py]: ah awesome, thanks! all looks good then, i'll merge and cherry pick

**Comment by chilir** [comment]
@daniellok-db thank you for following up on this! Sorry about disappearing, I got busy and then completely forgot about it 😅

**Pull request merged**
